cpp/src/arrow/python/udf.cc

```diff
@@ -0,0 +1,234 @@
+              "User-defined-functions are not supported for the datum kind ",
+              datum.ToString(batch[arg_id].kind()));
+      }
+    }
+    PyObject* result;
+    // result = PyObject_CallObject(function->obj(), arg_tuple);
+    result = cb(function->obj(), udf_context, arg_tuple);
+    RETURN_NOT_OK(CheckPyError());
+    if (result == Py_None) {
+      return Status::Invalid("Output is None, but expected an array");
+    }
+    // unwrapping the output for expected output type
+    if (is_scalar(result)) {
+      ARROW_ASSIGN_OR_RAISE(auto val, unwrap_scalar(result));
+      RETURN_NOT_OK(CheckOutputType(*output_type.type(), *val->type));
+      *out = Datum(val);
+      return Status::OK();
+    } else if (is_array(result)) {
+      ARROW_ASSIGN_OR_RAISE(auto val, unwrap_array(result));
+      RETURN_NOT_OK(CheckOutputType(*output_type.type(), *val->type()));
+      *out = Datum(val);
+      return Status::OK();
+    } else {
+      return Status::TypeError("Unexpected output type: ", Py_TYPE(result)->tp_name,
+                               " (expected Scalar or Array)");
+    }
+    return Status::OK();
+  }
+};
+
+// Status RegisterScalarFunction(PyObject* function, const ScalarUdfOptions& options) {
+//   if (function == nullptr) {
+//     return Status::Invalid("Python function cannot be null");
+//   }
+//   if (!PyCallable_Check(function)) {
+//     return Status::TypeError("Expected a callable Python object.");
+//   }
+//   auto doc = options.doc();
+//   auto arity = options.arity();
+//   auto in_types = options.input_types();
+//   auto exp_out_type = options.output_type();
+//   auto scalar_func =
+//       std::make_shared<compute::ScalarFunction>(options.name(), arity, std::move(doc));
+//   Py_INCREF(function);
+//   PythonUdf exec{std::make_shared<OwnedRefNoGIL>(function), std::move(exp_out_type)};
+//   compute::ScalarKernel kernel(
+//       compute::KernelSignature::Make(options.input_types(), options.output_type(),
+//                                      arity.is_varargs),
+//       std::move(exec));
+//   kernel.mem_allocation = compute::MemAllocation::NO_PREALLOCATE;
+//   kernel.null_handling = compute::NullHandling::COMPUTED_NO_PREALLOCATE;
+//   RETURN_NOT_OK(scalar_func->AddKernel(std::move(kernel)));
+//   auto registry = compute::GetFunctionRegistry();
+//   RETURN_NOT_OK(registry->AddFunction(std::move(scalar_func)));
+//   return Status::OK();
+// }
+
+Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback wrapper,
+                              const ScalarUdfOptions& options) {
+  if (user_function == nullptr) {
+    return Status::Invalid("Python function cannot be null");
+  }
+  if (!PyCallable_Check(user_function)) {
+    return Status::TypeError("Expected a callable Python object.");
+  }
+  auto doc = options.doc();
+  auto arity = options.arity();
+  auto in_types = options.input_types();
+  auto exp_out_type = options.output_type();
```

Comment:
nit, but is it really necessary to make copies of all these to just save a few characters over inlining them?